analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifierException.cs

```diff
@@ -0,0 +1,52 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.IO;
+using System.Text;
+using SonarAnalyzer.TestFramework.Verification.IssueValidation;
+
+namespace SonarAnalyzer.Test.TestFramework;
+
+public sealed class DiagnosticVerifierException : AssertFailedException
+{
+    private readonly string stackTrace;
+
+    public override string StackTrace =>
+        stackTrace;
+
+    internal DiagnosticVerifierException(List<VerificationMessage> messages) : base(messages.Select(x => x.FullDescription).JoinStr(Environment.NewLine))
+    {
+        if (messages.Count == 0)
+        {
+            throw new ArgumentException($"{nameof(messages)} cannot be empty", nameof(messages));
+        }
+        var builder = new StringBuilder();
+        foreach (var fileMessages in messages.Where(x => x.ShortDescription is not null && File.Exists(x.FullPath)).GroupBy(x => x.FullPath).OrderBy(x => x.Key))
```

Comment:
```csharp
        messages.Where(x => x.ShortDescription is not null && File.Exists(x.FullPath)).GroupBy(x => x.FullPath).OrderBy(x => x.Key)
```

I find this a bit overloaded - I'd separate it maybe in its own variable or to not iterate twice in a local function.

Reply:
What do you mean by "not iterate twice"?

I'll keep it, it's just below the edge to extract

Reply:
If you were to extract `var filtered =  messages.Where(x => x.ShortDescription is not null && File.Exists(x.FullPath)).GroupBy(x => x.FullPath).OrderBy(x => x.Key)` and then do `foreach (var message in filtered) ` you'd be iterating twice no?

Reply:
The `var filtered` doesn't evaluate the `messages` at all. It only creates a query for Where, GroupBy, OrderBy predicates.
Then, `foreach` will do first evaluation. There would need to be for example `.Count()` to iterate that twice.